deepchem/data/datasets.py

```diff
@@ -636,65 +637,127 @@ def iterbatches(self,
                   epoch=0,
                   deterministic=False,
                   pad_batches=False):
-    """Get an object that iterates over minibatches from the dataset.
-
+    """ Get an object that iterates over minibatches from the dataset. It is guaranteed
+    that the number of batches returned is math.ceil(len(dataset)/batch_size).
+    
     Each minibatch is returned as a tuple of four numpy arrays: (X, y, w, ids).
+
+
+    Parameters:
+    -----------
+    batch_size: int
+      Number of elements in a batch. If None, then it yields batches with size equal to the size
+      of each individual shard.
+
+    epoch: int
+      Not used
+
+    deterministic: bool
+      Whether or not we should should shuffle each shard before generating the batches.
+      Note that this is only local in the sense that it does not ever mix between different
+      shards.
+
+    pad_batches: bool
+      Whether or not we should pad the last batch, globally, such that it has exactly batch_size
+      elements.
+
+
     """
 
-    def iterate(dataset):
+    def iterate(dataset, batch_size):
       num_shards = dataset.get_number_shards()
       if not deterministic:
         shard_perm = np.random.permutation(num_shards)
       else:
         shard_perm = np.arange(num_shards)
-      pool = Pool(1)
+
+      # (ytz): Depending on the application, thread-based pools may be faster
+      # than process based pools, since process based pools need to pickle/serialize
+      # objects as an extra overhead. Also, as hideously as un-thread safe this looks,
+      # we're actually protected by the GIL.
+      pool = Pool(1)  # mp.dummy aliases ThreadPool to Pool
       next_shard = pool.apply_async(dataset.get_shard, (shard_perm[0],))
-      for i in range(num_shards):
+
+      total_yield = 0
+
+      if batch_size is None:
+        num_global_batches = num_shards
+      else:
+        num_global_batches = math.ceil(len(dataset) / batch_size)
```

Comment:
Note to self
```python
len(dataset) # literally reads in all the data
```

Reply:
We should implement some caching -- not part of this review.  I can store these things in the side car json file that I am making as part of the transition to a stable store.

Reply:
wait what? so I should revert to get_shape()[0][0] - afaik that just looks up the metadata